bmaptools/BmapHelpers.py

```diff
@@ -83,3 +96,54 @@ def program_is_available(name):
             return True
 
     return False
+
+
+def get_file_system_type(filepath):
+    """
+    Return the file system type the file object file resides on.
+    """
+
+    # normalize,resolve and get absolute path to the file object
+    f = os.path.realpath(filepath)
+    cmd = "df -T '%s' | awk 'NR==2 {print $2}'" % f
+    proc = subprocess.Popen(cmd, stdout=PIPE, stderr=PIPE, shell=True, universal_newlines=True)
```

Comment:
Is it possible to avoid running external programs for this? Less of this is generally better, less dependencies. Did you look at other solutions, like just scanning /proc/mounds, or using psutils, here is some helpful stackoverflow discussion: https://stackoverflow.com/questions/7718411/determine-device-of-filesystem-in-python?lq=1

Reply:
whilst psutil looks like a good approach, it looks like it comes with at a heavy price, the user would need to install gcc and python3-dev on linux and Xcode on mac.  Am i correct in thinking mac support is not a thing yet for bmap-tools and hence I dont need to worry about it?

Reply:
Probably we do not want psutil. I never worried about Windows or Mac.

I guess 'df' is OK, but please, do not pipe to awk, python has enough tools for parsing strings. I think all you need will be output.split()[0].

Also, please handle errors. Check how / if they are handled in different places in this project. I went through evolution of how I handle exception in my programs, and I do not remember how I did it in this project, but just be consistent with the rest of the code.

Reply:
if mac support is required then, "df -T '%s' | awk 'NR==2 {print $2}" can be replaced with a single call to "mount" and the output of this would need to be parsed and correlated with a python function to determine the mount point that the image file resides on

Reply:
> `cmd = "df -T '%s' | awk 'NR==2 {print $2}'" % f`

This is vulnerable to [shell injection](https://xkcd.com/327/). If the filename is `'$(rm -fr /)` (including the single quote, to escape from the single quotes in the format string) then running bmaptool will remove all your files. In general, avoid using `shell=True` with `subprocess`.

The best and most direct way to get this information would be `os.statvfs(f).f_type` and comparing with the filesystem's magic number, but Python doesn't seem to implement that struct field...

Using `subprocess.Popen(['stat', '-f', '--printf=%t'], ...)` (without `shell=True`), and comparing with the magic number of a ZFS filesystem (possibly 0x2f52f5?), might be a less bad way to do this with external commands; or maybe `%T` instead of `%t` and comparing with `zfs`, or similar.

Reply:
looks a better way to go

Reply:
a quick test, however is resulting in stat output not accurately  identifying the filesystem type being used. testing on files on an ext4 file system results in ext2/ext3 being returned for stat whereas df returns ext4.

Reply:
ext2, ext3 and ext4 all have the same magic number, so it's expected that different tools will label them differently. What matters for this PR is surely whether ZFS can be detected?

Reply:
given that these are support helper functions, I was aiming for this particular function to be accurate and reusable in-case the file system needs to be known for some other reasons in the future.

Reply:
shell=true should no longer be needed with the removal of the pipe as per @dedekind request.

Reply:
taking the best of @smcv  @dedekind recommendations it should be possible to do something like:

subprocess.Popen(['df', '-T', '--', f], ...)

and the process the output in python